### PR TITLE
Added and Changed docstrings to make it readable.

### PR DESCRIPTION
**Pull request by ranbir7**

### CODE REVIEW

File: autoprof/AP_config.py [hunk 1]
```diff
@@ -49,5 +52,5 @@ def set_logging_output(stdout = True, filename = None, **kwargs):
         out_handler.setLevel(kwargs.get("filename_level", logging.INFO))
         out_handler.setFormatter(kwargs.get("filename_formatter", logging.Formatter("%(asctime)s:%(levelname)s: %(message)s")))
         ap_logger.addHandler(out_handler)
-        ap_logger.debug("logging now going to stdout")
+        ap_logger.debug("logging now going to %s" % filename)
```
Comment (ranbir7) [inline autoprof/AP_config.py:55]: Hey @ConnorStone628 
I changed this to Filename 
since it was in the filename if-block.
If there's any changes to be made, Let me know.

**Comment by ranbir7** [comment]
I'll complete atleast 4 files and then I'll make it ready for review
till then just mention my mistakes
I'm trying my best to provide error free docs and code.
Hope u like it !

File: autoprof/AP_config.py [hunk 2]
```diff
@@ -15,19 +15,22 @@
...
-        stdout_level: the logging level of messages written to stdout, this can be different from the file level. default: logging.INFO
-        stdout_formatter: a logging.Formatter object which determines what information to include with the logging message only when printing to stdout. default: "%(message)s"
-        filename_level: the logging level of messages written to the log file, this can be different from the stdout level. default: logging.INFO
-        filename_formatter: a logging.Formatter object which determines what information to include with the logging message only when printing to the log file. default: "%(asctime)s:%(levelname)s: %(message)s"
-
+        stdout (bool): If True, log messages will be printed to the console (standard output). Default is True.
+        filename (str): If given as a string, this will be the name of the file that log messages are written to. 
+                        If None, no logging file will be used. Default is None.
+        stdout_level (int): The logging level of messages written to stdout, this can be different from the file level.
```
Comment (ConnorStoneAstro) [inline autoprof/AP_config.py]: The type for this is not `int` it is a logging level object. typically `logging.DEBUG, logging.INFO, logging.WARNING`

File: autoprof/AP_config.py (hunk 1 above)
Comment (ConnorStoneAstro) [reply autoprof/AP_config.py:55]: Good catch!

File: autoprof/fit/__init__.py [hunk 3]
```diff
@@ -3,3 +3,26 @@
 from .gradient import *
 from .langevin import *
 from .iterative import *
+"""
+base: This module defines the base class AutoprofBase,
```
Comment (ConnorStoneAstro) [inline autoprof/fit/__init__.py]: instead of `AutoprofBase` it is `BaseOptimizer`

File: autoprof/fit/__init__.py [hunk 4]
```diff
@@ -3,3 +3,26 @@
 from .gradient import *
 from .langevin import *
 from .iterative import *
+"""
+base: This module defines the base class AutoprofBase, 
+      which is used as the parent class for all profiling algorithms in AutoProf.
```
Comment (ConnorStoneAstro) [inline autoprof/fit/__init__.py]: instead of `profiling` it should be `optimization` throughout the comment

File: autoprof/fit/__init__.py [hunk 5]
```diff
@@ -3,3 +3,26 @@
 from .gradient import *
 from .langevin import *
 from .iterative import *
+"""
+base: This module defines the base class AutoprofBase, 
+      which is used as the parent class for all profiling algorithms in AutoProf. 
+      This module contains helper functions used across multiple profiling algorithms, 
+      such as computing gradients and making copies of models.
+
+lm: This module defines the class LMProfiler,
```
Comment (ConnorStoneAstro) [inline autoprof/fit/__init__.py]: the lm class is just `LM` right now, not `LMProfiler`

File: autoprof/fit/__init__.py [hunk 6]
```diff
@@ -3,3 +3,26 @@
...
+base: This module defines the base class AutoprofBase, 
+      which is used as the parent class for all profiling algorithms in AutoProf. 
+      This module contains helper functions used across multiple profiling algorithms, 
+      such as computing gradients and making copies of models.
+
+lm: This module defines the class LMProfiler, 
+    which uses the Levenberg-Marquardt algorithm to perform profiling. 
+    This algorithm adjusts the learning rate at each step to find the optimal value.
+
+gradient: This module defines the class GradientProfiler,
```
Comment (ConnorStoneAstro) [inline autoprof/fit/__init__.py]: the gradient optimizer class is just `Grad`

File: autoprof/fit/__init__.py [hunk 7]
```diff
@@ -3,3 +3,26 @@
...
+
+lm: This module defines the class LMProfiler, 
+    which uses the Levenberg-Marquardt algorithm to perform profiling. 
+    This algorithm adjusts the learning rate at each step to find the optimal value.
+
+gradient: This module defines the class GradientProfiler, 
+          which uses a simple gradient descent algorithm to perform profiling. 
+          This algorithm adjusts the learning rate at each step to find the optimal value.
+
+langevin: This module defines the class Langevin,
```
Comment (ConnorStoneAstro) [inline autoprof/fit/__init__.py]: The langevin class is being depreciated, I am currently working on a new one. Still the class in the langevin module is called `MALA`

File: autoprof/fit/__init__.py [hunk 8]
```diff
@@ -3,3 +3,26 @@
...
+
+gradient: This module defines the class GradientProfiler, 
+          which uses a simple gradient descent algorithm to perform profiling. 
+          This algorithm adjusts the learning rate at each step to find the optimal value.
+
+langevin: This module defines the class Langevin, 
+          which uses the Langevin dynamics algorithm to perform profiling. 
+          This algorithm simulates a physical system to find the optimal value.
+
+iterative: This module defines the class iter,
```
Comment (ConnorStoneAstro) [inline autoprof/fit/__init__.py]: this is right, except the first letter is capitalized: `Iter`

File: autoprof/fit/__init__.py [hunk 9]
```diff
@@ -3,3 +3,26 @@
...
+          which uses a simple gradient descent algorithm to perform profiling. 
+          This algorithm adjusts the learning rate at each step to find the optimal value.
+
+langevin: This module defines the class Langevin, 
+          which uses the Langevin dynamics algorithm to perform profiling. 
+          This algorithm simulates a physical system to find the optimal value.
+
+iterative: This module defines the class iter, 
+            which uses an iterative algorithm to perform profiling. 
+            This algorithm repeatedly trains the model with different learning rates until it finds the optimal value.
```
Comment (ConnorStoneAstro) [inline autoprof/fit/__init__.py]: the algorithm repeatedly fits each model individually until they all converge. This uses less memory at any one point

File: autoprof/fit/langevin.py [hunk 10]
```diff
@@ -2,6 +2,7 @@
 import torch
 import numpy as np
 from time import time
+from typing import Union, Dict, Optional
```
Comment (ConnorStoneAstro) [inline autoprof/fit/langevin.py:5]: the typing import should come first

File: autoprof/fit/langevin.py [hunk 11]
```diff
@@ -51,7 +52,24 @@ class MALA(BaseOptimizer):
 
     Borrowed heavily from Alexandre Adam: https://github.com/AlexandreAdam/
     """
-    def __init__(self, model, initial_state = None, **kwargs):
+    def __init__(self, model: 'AutoProf', initial_state: Optional[Dict[str,torch.Tensor]] = None, **kwargs)-> None:
```
Comment (ConnorStoneAstro) [inline autoprof/fit/langevin.py:55]: The model is  of type `AutoProf_Model`, The initial state should more likely be: `Optional[Sequence]`

File: autoprof/fit/langevin.py [hunk 12]
```diff
@@ -51,7 +52,24 @@ class MALA(BaseOptimizer):
...
+        """
+    Initialize a Langevin Dynamics sampler object for AutoProf.
+
+    Args:
+        model (AutoProf): An AutoProf object representing the model to be sampled.
```
Comment (ConnorStoneAstro) [inline autoprof/fit/langevin.py:60]: type should be `AutoProf_Model`

File: autoprof/fit/langevin.py [hunk 13]
```diff
@@ -51,7 +52,24 @@ class MALA(BaseOptimizer):
...
+        initial_state (dict, optional): A dictionary containing the initial values of the model parameters to be sampled.
```
Comment (ConnorStoneAstro) [inline autoprof/fit/langevin.py:61]: initial state can be any sequence

File: autoprof/fit/langevin.py [hunk 14]
```diff
@@ -107,37 +134,56 @@ def accept(self, log_alpha):
...
         Computes the log probability difference between the proposed state x' and the current state x
             log p(x') - log p(x)
         """
         v = proposed_state - current_state
         T = self.delta_logp_steps + 1
         D = current_state.shape
+        return simps(self.integrand, 0., 1., self.delta_logp_steps, device=self.model.device)
 
-        def integrand(t):
+    def integrand(self, t: torch.Tensor)-> torch.Tensor:
```
Comment (ConnorStoneAstro) [inline autoprof/fit/langevin.py:147]: The `integrand` function is intentionally defined inside the the `delta_logp` function for simplicity. I don't think this needs to change

**Review by ConnorStoneAstro — COMMENTED** [review]
Lots of great stuff, I've added some comments to address

File: autoprof/fit/langevin.py (hunk 14 above)
Comment (ranbir7) [reply autoprof/fit/langevin.py:147]: Okay I'll revert it back...

**Review by ConnorStoneAstro — APPROVED** [review]
Looks ready

**Comment by ranbir7** [comment]
wait there's still more
I'll solve every thing u have addressed 
Then I'll Make it ready for review 
alright?

**Comment by ConnorStoneAstro** [comment]
Ok, but you can ignore the comments on the `langevin.py` file. That one is being replaced soon.
